pkg/server/commands/listusers/list_users_rpc_test.go

```diff
@@ -2,301 +2,1132 @@ package listusers
+				UserFilters: []*openfgav1.ListUsersFilter{
+					{
+						Type: "user",
+					},
+				},
+			},
+			model: model,
 			tuples: []*openfgav1.TupleKey{
-				tuple.NewTupleKey("document:1", "editor", "user:jon"),
+				tuple.NewTupleKey("document:1", "parent", "folder:x"),
+				tuple.NewTupleKey("folder:x", "viewer", "user:maria"),
+				tuple.NewTupleKey("folder:no-doc", "viewer", "user:maria"),
+				tuple.NewTupleKey("document:1", "parent", "folder:no-user"),
 			},
-			expectedUsers: []*openfgav1.User{
-				{
-					User: &openfgav1.User_Object{
-						Object: &openfgav1.Object{Type: "user", Id: "jon"},
+			expectedUsers: []string{"user:maria"},
+		},
+		{
+			name:                  "ttu_folder_granularity",
+			TemporarilySkipReason: "because returns `folder:x` despite impossible edge folder=>document#viewer",
+			req: &openfgav1.ListUsersRequest{
+				Object:   &openfgav1.Object{Type: "document", Id: "1"},
+				Relation: "viewer",
+				UserFilters: []*openfgav1.ListUsersFilter{
+					{
+						Type: "folder",
 					},
 				},
 			},
+			model: model,
+			tuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("document:1", "parent", "folder:x"),
+				tuple.NewTupleKey("folder:x", "viewer", "user:maria"),
+			},
+			expectedUsers: []string{},
 		},
 		{
-			name: "rewritten_direct_relationship_through_ttu",
+			name:                  "ttu_with_computed_relation_user_granularity",
+			TemporarilySkipReason: "because results not deduplicated",
 			req: &openfgav1.ListUsersRequest{
-				StoreId:              storeID,
-				AuthorizationModelId: modelID,
-				Object:               &openfgav1.Object{Type: "document", Id: "1"},
-				Relation:             "viewer",
+				Object:   &openfgav1.Object{Type: "document", Id: "1"},
+				Relation: "viewer",
 				UserFilters: []*openfgav1.ListUsersFilter{
 					{
 						Type: "user",
 					},
 				},
 			},
-			typeDefinitions: parser.MustParse(`
+			model: `model
+			schema 1.1
+		  type user
+		
+		  type folder
+			relations
+				define owner: [user]
+				define editor: [user] or owner
+				define viewer: [user] or owner or editor
+				define unrelated_not_computed: [user]
+		
+		  type document
+			relations
+			  define parent: [folder]
+			  define viewer: viewer from parent`,
+
+			tuples: []*openfgav1.TupleKey{
+				tuple.NewTupleKey("document:1", "parent", "folder:x"),
+				tuple.NewTupleKey("folder:x", "viewer", "user:maria"),
+				tuple.NewTupleKey("folder:x", "editor", "user:will"),
+				tuple.NewTupleKey("folder:x", "owner", "user:jon"),
+				tuple.NewTupleKey("folder:x", "unrelated_not_computed", "user:poovam"),
+
+				tuple.NewTupleKey("folder:no-doc", "viewer", "user:maria"),
+				tuple.NewTupleKey("document:1", "parent", "folder:no-user"),
+			},
+			expectedUsers: []string{"user:maria", "user:will", "user:jon"},
+		},
+	}
+
+	tests.runListUsersTestCases(t)
+}
+
+func TestListUsersCycles(t *testing.T) {
+	t.Cleanup(func() {
+		goleak.VerifyNone(t)
+	})
+	tests := ListUsersTests{
```

Comment:
It would also be good to add one more that establishes a cycle through a TTU hierarchy. e.g.

```
type user

type folder
  relations
    define viewer: [user, document#viewer]

type document
  relations
    define parent: [folder]
    define viewer: viewer from parent
```

- document:1#parent@folder:x
- folder:x#viewer@document:1#viewer

ListUsers(document:1#viewer, user_filters=[{"type": "user"}])